packages/relay/src/services/singletons/UnirepSocialSynchronizer.ts

```diff
@@ -225,15 +233,68 @@ export class UnirepSocialSynchronizer extends Synchronizer {
             epoch: epoch,
         })
 
-        // if there's no data in EpochKeyAction then do nothing
-        if (rows == 0) return result
-
         // make sure all data whose epochs are before the current ended one is deleted as well
-        db.delete('EpochKeyAction', {
+        if (rows) {
+            db.delete('EpochKeyAction', {
+                where: {
+                    epoch: { lte: epoch },
+                },
+            })
+        }
+
+        // Settle reports
+        const votingReports = await this.db.findMany('ReportHistory', {
             where: {
-                epoch: { lte: epoch },
+                status: ReportStatus.VOTING,
             },
         })
+        // Go through all the voting reports
+        for (const report of votingReports) {
+            const adjudicators = report.adjudicatorsNullifier || []
+            const agreeVotes = adjudicators.filter(
+                (adj: Adjudicator) =>
+                    adj.adjudicateValue === AdjudicateValue.AGREE
+            ).length
+            const disagreeVotes = adjudicators.filter(
+                (adj: Adjudicator) =>
+                    adj.adjudicateValue === AdjudicateValue.DISAGREE
+            ).length
+            // If the current epoch > reportEpoch AND sum of votes > threshold AND vote value > 0
+            if (
+                epoch >= report.reportEpoch &&
+                report.adjudicateCount > REPORT_SETTLE_VOTE_THRESHOLD &&
+                agreeVotes !== disagreeVotes
+            ) {
+                // Then update the status of the report to WaitingForTx
+                db.update('ReportHistory', {
+                    where: { reportId: report.reportId },
+                    update: {
+                        status: ReportStatus.WAITING_FOR_TRANSACTION,
+                    },
+                })
+                // Then update the status of the post or comment to Onchain or Disagreed
+                const newStatus =
```

Comment:
Ternary operator is not easy to read in this case, we can do a tricky way to modify the order of comment status to below. After that, the `ON_CHAIN` & `DISAGREED` enum become same number on comment status and post status.
```typescript
export enum CommentStatus {
    NOT_ON_CHAIN,
    ON_CHAIN,
    REPORTED,
    DISAGREED,
    DELETED,
}
```
Then update status logic can be modified to
```typescript
// 1: ON_CHAIN, 3: DISAGREED
const newStatus = agreeVotes > disagreeVotes ? 1 : 3
const tableName = report.type == ReportType.POST ? 'Post' : 'Comment'

this.db.update(tableName, {
    where: {
        // concat to PostId / CommentId
        [`${tableName.toLowerCase()}Id`]: report.objectId,
    },
    update: {
        newStatus
    }
})
```